From 59c407cf717ab5e4c639b6c466ca426a0434ad07 Mon Sep 17 00:00:00 2001
From: Henry Chen <chenx97@aosc.io>
Date: Thu, 4 Jan 2024 00:05:49 -0500
Subject: [PATCH 23/62] hid-lenovo: detect false-positives

This patch allows the workaround to work again when wheel signals
are detected after identifying the keyboard as bug-free. False
positives happen when the middle button is released while the
kernel doesn't see it.
---
 drivers/hid/hid-lenovo.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/hid/hid-lenovo.c b/drivers/hid/hid-lenovo.c
index 149a3c74346b..2db6f85cb4a1 100644
--- a/drivers/hid/hid-lenovo.c
+++ b/drivers/hid/hid-lenovo.c
@@ -686,7 +686,8 @@ static int lenovo_event_cptkbd(struct hid_device *hdev,
 {
 	struct lenovo_drvdata *cptkbd_data = hid_get_drvdata(hdev);
 
-	if (cptkbd_data->middlebutton_state != 3) {
+	if (cptkbd_data->middlebutton_state != 3 ||
+				(usage->code == REL_WHEEL || usage->code == REL_HWHEEL)) {
 		/* REL_X and REL_Y events during middle button pressed
 		 * are only possible on patched, bug-free firmware
 		 * so set middlebutton_state to 3
-- 
2.44.0

